docs/sources/userguide/plugins.md

```diff
@@ -0,0 +1,139 @@
+Docker Plugins MVP
+==================
+
+Plugins provide a mechanism for hooking into the Docker engine to extend its behavior.
+
+HTTP over a local UNIX socket is used to communicate between Docker and its plugins.
+
+Plugins are distributed as containers.
+
+# Usage
+
+Plugins can be loaded using a special `docker plugin load` command, as follows:
+
+```
+$ docker plugin load clusterhq/flocker-plugin
+[... fetching...]
+Plugin flocker:v0.1 loaded and registered with extension points:
+* volumes
+Plugin flocker activated.
+```
+
+This is really just syntactic sugar for the following:
+
+```
+$ docker run -d -e \
+	-v /var/run/docker-plugins/flocker.sock:/var/run/plugin.sock \
+	clusterhq/flocker-plugin
+```
+
+Docker then waits for the plugin to start listening on the socket (it polls the socket until it gets a successful response to an HTTP query to `/v1/plugin/handshake` on the socket, which returns with just a list of subsystems the plugin is interested in `["volume"]`). According to the type of plugin which is negotiated in the handshake, Docker registers the plugin to send it HTTP requests on certain events.
+
+Other `plugin` subcommands can include `list`, `status`, `upgrade`, and `unload`.
+
+# Types of plugin
+
+Each plugin type defines a straightforward (MVP) protocol.
+
+## Volume
+
+The simplest of plugin types, `volume` provides a single request-response type:
+
+**Request**
+
+```
+{DockerVolumesExtensionVersion: 1,
+ Action: "create",
+ HostPath: "/path",
+ ContainerID: "abcdef123"}
+```
+
+**Response**
+
+```
+{DockerVolumesExtensionVersion: 1,
+ ModifiedHostPath: "/newpath"} 
+```
+
+In the initial version, if the plugin responds with an empty string (`""`) as the `ModifiedHostPath`, the response is ignored.
+
+See reference implementation:
+
+* [Docker volumes extension mechanism](https://github.com/cpuguy83/docker/compare/ddb366ee9a07e3feab766cc712c9683ad0c3c309)
+* [Flocker extension](https://github.com/clusterhq/powerstrip-flocker/compare/docker-volume-extension)
+
+**Demo**
+
+Despite the name, this demo does not use powerstrip!
+
+```
+$ git clone https://github.com/clusterhq/powerstrip-flocker
+$ cd powerstrip-flocker/vagrant-aws
+$ git checkout docker-volume-extension
+$ vagrant up --provider=aws
+$ vagrant ssh node1
+node1$ docker run -v /flocker/test:/data ubuntu sh -c "echo fish > /data/file.txt"
+node1$ exit
+$ vagrant ssh node2
+node2$ docker run -v /flocker/test:/data ubuntu cat /data/file.txt
+fish
+```
+
+In future, we will want to extend this to notify volume extensions when containers start and stop, so that the underlying volumes mechanism can hold and release *leases* on volumes.
+
+### Volumes CLI
+
+We may want to extend the volumes CLI in due course, so you can do something like the following:
+
+```
+$ docker volume create --driver flocker \
+	--metadata size=50G,tier=ssd,replication=global nicevol
```

Comment:
This should match the --env/--volume semantics:

`--metadata size=50G --metadata tier=ssd --metadata replication=global`